In [1]:
# Import dependencies
import pandas as pd
from google.colab import files

In [2]:
# Load in the Migration_Population Dataset
path = "https://github.com/pavlarsen/Final_Project/blob/main/Resources/Data/Migrant_Population_2.xlsx?raw=true"
migrant_population_df = pd.read_excel(path, sheet_name = 1, skiprows = [0, 1,2,3,4,5,6,7,8], header = 1)
migrant_population_df.head(10)

Index  Year Region, development group, country or area Notes  \
0      1  1990                                      WORLD   NaN   
1      2  1995                                      WORLD   NaN   
2      3  2000                                      WORLD   NaN   
3      4  2005                                      WORLD   NaN   
4      5  2010                                      WORLD   NaN   
5      6  2015                                      WORLD   NaN   
6      7  2020                                      WORLD   NaN   
7      8  1990                         Sub-Saharan Africa   NaN   
8      9  1995                         Sub-Saharan Africa   NaN   
9     10  2000                         Sub-Saharan Africa   NaN   

   Location code Type of data      0-4      5-9     10-14     15-19  ...  \
0            900          NaN  4940094  6508813   7870896   9631753  ...   
1            900          NaN  4683966  6256492   7690520   9453166  ...   
2            900          NaN  4726060  6151369   7603830   9578087  ...   
3            900          NaN  4912830  6562128   8100234  10652190  ...   
4            900          NaN  5772791  7338510   8458240  11057021  ...   
5            900          NaN  6989696  8088701   8615587  10947107  ...   
6            900          NaN  8226578  9969159  10212634  12529837  ...   
7            947          NaN  1083361  1138728   1130403   1290880  ...   
8            947          NaN  1012625  1158292   1199559   1358836  ...   
9            947          NaN   924164   990600   1023440   1202364  ...   

    35-39.2   40-44.2   45-49.2  50-54.2  55-59.2  60-64.2  65-69.2  70-74.2  \
0   6866805   6216888   5213839  4428584  3978529  3691290  2973005  2548763   
1   7548080   6816854   5911764  4995075  4287775  3823263  3119032  2864763   
2   8255608   7540512   6559199  5758957  4686452  4172300  3326001  3119934   
3   8907407   8233674   7377669  6499130  5514576  4567244  3706760  3278077   
4  10181838   9371840   8598912  7543434  6445198  5557592  4217541  3623915   
5  11705637  10766564   9629761  8636848  7447729  6284680  5096549  4004130   
6  13442369  12331582  10864220  9485189  8299468  7038905  5745085  4525479   
7    460870    348158    262423   209807   159495   141173    92568    57315   
8    522964    403097    310397   247480   192278   158479   111530    69302   
9    502941    385269    296160   229043   182921   145742   102447    67414   

     75+.2    Total.2  
0  4928529   75422690  
1  5080075   79708475  
2  5690722   85510752  
3  5868837   93783747  
4  6527536  107042306  
5  7282795  119720586  
6  8752330  134942261  
7    80052    6237924  
8    83101    6761344  
9    78441    6211237  

[10 rows x 57 columns]

In [3]:
# Rename the columns we will use
migrant_population_df.rename(columns={"Region, development group, country or area": "Country_Name", "Location code": "Country_ID", "Total":"Total_Migrant_Population", "Total.1":"Migrant_Male_Population","Total.2":"Migrant_Female_Population"}, inplace = True)

In [4]:
# Check the previous code
migrant_population_df.head()

Index  Year Country_Name Notes  Country_ID Type of data      0-4      5-9  \
0      1  1990        WORLD   NaN         900          NaN  4940094  6508813   
1      2  1995        WORLD   NaN         900          NaN  4683966  6256492   
2      3  2000        WORLD   NaN         900          NaN  4726060  6151369   
3      4  2005        WORLD   NaN         900          NaN  4912830  6562128   
4      5  2010        WORLD   NaN         900          NaN  5772791  7338510   

     10-14     15-19  ...   35-39.2  40-44.2  45-49.2  50-54.2  55-59.2  \
0  7870896   9631753  ...   6866805  6216888  5213839  4428584  3978529   
1  7690520   9453166  ...   7548080  6816854  5911764  4995075  4287775   
2  7603830   9578087  ...   8255608  7540512  6559199  5758957  4686452   
3  8100234  10652190  ...   8907407  8233674  7377669  6499130  5514576   
4  8458240  11057021  ...  10181838  9371840  8598912  7543434  6445198   

   60-64.2  65-69.2  70-74.2    75+.2  Migrant_Female_Population  
0  3691290  2973005  2548763  4928529                   75422690  
1  3823263  3119032  2864763  5080075                   79708475  
2  4172300  3326001  3119934  5690722                   85510752  
3  4567244  3706760  3278077  5868837                   93783747  
4  5557592  4217541  3623915  6527536                  107042306  

[5 rows x 57 columns]

In [5]:
# Create a reduced_df that only has the columns we will use
reduced_df = migrant_population_df[["Country_ID", "Country_Name", "Year", "Total_Migrant_Population", "Migrant_Male_Population", "Migrant_Female_Population"]]
reduced_df.head()

Country_ID Country_Name  Year  Total_Migrant_Population  \
0         900        WORLD  1990                 152986157   
1         900        WORLD  1995                 161289976   
2         900        WORLD  2000                 173230585   
3         900        WORLD  2005                 191446828   
4         900        WORLD  2010                 220983187   

   Migrant_Male_Population  Migrant_Female_Population  
0                 77563467                   75422690  
1                 81581501                   79708475  
2                 87719833                   85510752  
3                 97663081                   93783747  
4                113940881                  107042306

In [6]:
# Get a list of country codes in order to drop the regions that will not be part of the dataset
code_path = "https://raw.githubusercontent.com/lukes/ISO-3166-Countries-with-Regional-Codes/master/all/all.csv"
code_df = pd.read_csv(code_path, encoding = "UTF-8", header = 0)
country_code = code_df["country-code"].values.tolist()
country_code

[4,
 248,
 8,
 12,
 16,
 20,
 24,
 660,
 10,
 28,
 32,
 51,
 533,
 36,
 40,
 31,
 44,
 48,
 50,
 52,
 112,
 56,
 84,
 204,
 60,
 64,
 68,
 535,
 70,
 72,
 74,
 76,
 86,
 96,
 100,
 854,
 108,
 132,
 116,
 120,
 124,
 136,
 140,
 148,
 152,
 156,
 162,
 166,
 170,
 174,
 178,
 180,
 184,
 188,
 384,
 191,
 192,
 531,
 196,
 203,
 208,
 262,
 212,
 214,
 218,
 818,
 222,
 226,
 232,
 233,
 748,
 231,
 238,
 234,
 242,
 246,
 250,
 254,
 258,
 260,
 266,
 270,
 268,
 276,
 288,
 292,
 300,
 304,
 308,
 312,
 316,
 320,
 831,
 324,
 624,
 328,
 332,
 334,
 336,
 340,
 344,
 348,
 352,
 356,
 360,
 364,
 368,
 372,
 833,
 376,
 380,
 388,
 392,
 832,
 400,
 398,
 404,
 296,
 408,
 410,
 414,
 417,
 418,
 428,
 422,
 426,
 430,
 434,
 438,
 440,
 442,
 446,
 450,
 454,
 458,
 462,
 466,
 470,
 584,
 474,
 478,
 480,
 175,
 484,
 583,
 498,
 492,
 496,
 499,
 500,
 504,
 508,
 104,
 516,
 520,
 524,
 528,
 540,
 554,
 558,
 562,
 566,
 570,
 574,
 807,
 580,
 578,
 512,
 586,
 585,
 275,
 591

In [7]:
# Drop the columns that refer to regions (we only want the data for countries)
country_migration_df = reduced_df[migrant_population_df["Country_ID"].isin(country_code)]
country_migration_df.head()

Country_ID Country_Name  Year  Total_Migrant_Population  \
161         108      Burundi  1990                    333110   
162         108      Burundi  1995                    254853   
163         108      Burundi  2000                    125628   
164         108      Burundi  2005                    192774   
165         108      Burundi  2010                    247321   

     Migrant_Male_Population  Migrant_Female_Population  
161                   163267                     169843  
162                   124165                     130688  
163                    61094                      64534  
164                    94413                      98361  
165                   121762                     125559

In [8]:
# Join the tables and drop the columns that won't be used
merged_df = pd.merge(country_migration_df, code_df, left_on = "Country_ID", right_on = "country-code", how="left")
country_code_df = merged_df[["Country_ID", "alpha-2", "Country_Name", "Year", "Total_Migrant_Population", "Migrant_Male_Population", "Migrant_Female_Population"]]

In [9]:
# Check the DataFrame
country_code_df.head()

Country_ID alpha-2 Country_Name  Year  Total_Migrant_Population  \
0         108      BI      Burundi  1990                    333110   
1         108      BI      Burundi  1995                    254853   
2         108      BI      Burundi  2000                    125628   
3         108      BI      Burundi  2005                    192774   
4         108      BI      Burundi  2010                    247321   

   Migrant_Male_Population  Migrant_Female_Population  
0                   163267                     169843  
1                   124165                     130688  
2                    61094                      64534  
3                    94413                      98361  
4                   121762                     125559

In [10]:
# Change the name column for the "alpha-2" column
renamed_df = country_code_df.rename(columns={"alpha-2":"Country_ID_Alpha"})

In [11]:
# Check for null values()
renamed_df.isna().sum()

Country_ID                   0
Country_ID_Alpha             7
Country_Name                 0
Year                         0
Total_Migrant_Population     0
Migrant_Male_Population      0
Migrant_Female_Population    0
dtype: int64

In [12]:
# Identify the rows that have NaN values.
renamed_df[renamed_df["Country_ID_Alpha"].isna()]

Country_ID Country_ID_Alpha Country_Name  Year  Total_Migrant_Population  \
273         516              NaN      Namibia  1990                    120641   
274         516              NaN      Namibia  1995                    115372   
275         516              NaN      Namibia  2000                    135547   
276         516              NaN      Namibia  2005                    107347   
277         516              NaN      Namibia  2010                    103826   
278         516              NaN      Namibia  2015                    101618   
279         516              NaN      Namibia  2020                    109391   

     Migrant_Male_Population  Migrant_Female_Population  
273                    63748                      56893  
274                    61310                      54062  
275                    73154                      62393  
276                    57625                      49722  
277                    55803                      48023  
278                    54739                      46879  
279                    59024                      50367

In [13]:
# Replace the nan values with the country code for Namibia ("NA")
complete_df = renamed_df.fillna("NA")
complete_df.isna().sum()

Country_ID                   0
Country_ID_Alpha             0
Country_Name                 0
Year                         0
Total_Migrant_Population     0
Migrant_Male_Population      0
Migrant_Female_Population    0
dtype: int64

In [14]:
# Check that the dataset has the correct data types
complete_df.dtypes

Country_ID                    int64
Country_ID_Alpha             object
Country_Name                 object
Year                          int64
Total_Migrant_Population      int64
Migrant_Male_Population       int64
Migrant_Female_Population     int64
dtype: object

In [15]:
complete_df.head()

Country_ID Country_ID_Alpha Country_Name  Year  Total_Migrant_Population  \
0         108               BI      Burundi  1990                    333110   
1         108               BI      Burundi  1995                    254853   
2         108               BI      Burundi  2000                    125628   
3         108               BI      Burundi  2005                    192774   
4         108               BI      Burundi  2010                    247321   

   Migrant_Male_Population  Migrant_Female_Population  
0                   163267                     169843  
1                   124165                     130688  
2                    61094                      64534  
3                    94413                      98361  
4                   121762                     125559

In [16]:
# Count rows
complete_df["Country_ID"].count()

1638

In [17]:
# Check for empty values
complete_df.loc[complete_df["Migrant_Male_Population"] == 0]

Country_ID Country_ID_Alpha                          Country_Name  Year  \
105          728               SS                           South Sudan  1990   
106          728               SS                           South Sudan  1995   
107          728               SS                           South Sudan  2000   
108          728               SS                           South Sudan  2005   
462          158               TW      China, Taiwan Province of China*  1990   
463          158               TW      China, Taiwan Province of China*  1995   
464          158               TW      China, Taiwan Province of China*  2000   
465          158               TW      China, Taiwan Province of China*  2005   
466          158               TW      China, Taiwan Province of China*  2010   
467          158               TW      China, Taiwan Province of China*  2015   
468          158               TW      China, Taiwan Province of China*  2020   
980          499               ME                            Montenegro  1990   
981          499               ME                            Montenegro  1995   
982          499               ME                            Montenegro  2000   
983          499               ME                            Montenegro  2005   
1127         535               BQ     Bonaire, Sint Eustatius and Saba*  1990   
1128         535               BQ     Bonaire, Sint Eustatius and Saba*  1995   
1129         535               BQ     Bonaire, Sint Eustatius and Saba*  2000   
1130         535               BQ     Bonaire, Sint Eustatius and Saba*  2005   
1131         535               BQ     Bonaire, Sint Eustatius and Saba*  2010   
1132         535               BQ     Bonaire, Sint Eustatius and Saba*  2015   
1133         535               BQ     Bonaire, Sint Eustatius and Saba*  2020   
1155         531               CW                              Curaçao*  1990   
1156         531               CW                              Curaçao*  1995   
1157         531               CW                              Curaçao*  2000   
1158         531               CW                              Curaçao*  2005   
1225         652               BL                     Saint Barthélemy*  1990   
1226         652               BL                     Saint Barthélemy*  1995   
1227         652               BL                     Saint Barthélemy*  2000   
1228         652               BL                     Saint Barthélemy*  2005   
1229         652               BL                     Saint Barthélemy*  2010   
1230         652               BL                     Saint Barthélemy*  2015   
1231         652               BL                     Saint Barthélemy*  2020   
1246         663               MF           Saint Martin (French part)*  1990   
1247         663               MF           Saint Martin (French part)*  1995   
1248         663               MF           Saint Martin (French part)*  2000   
1249         663               MF           Saint Martin (French part)*  2005   
1250         663               MF           Saint Martin (French part)*  2010   
1251         663               MF           Saint Martin (French part)*  2015   
1252         663               MF           Saint Martin (French part)*  2020   
1260         534               SX            Sint Maarten (Dutch part)*  1990   
1261         534               SX            Sint Maarten (Dutch part)*  1995   
1262         534               SX            Sint Maarten (Dutch part)*  2000   

      Total_Migrant_Population  Migrant_Male_Population  \
105                          0                        0   
106                          0                        0   
107                          0                        0   
108                          0                        0   
462                          0                        0   
463                          0                        0   
464                         

In [19]:
# Count the empty values
complete_df.loc[(complete_df["Migrant_Male_Population"] == 0) | (complete_df["Migrant_Female_Population"] == 0) | (complete_df["Total_Migrant_Population"] == 0)].count()

Country_ID                   43
Country_ID_Alpha             43
Country_Name                 43
Year                         43
Total_Migrant_Population     43
Migrant_Male_Population      43
Migrant_Female_Population    43
dtype: int64

In [20]:
# Make a copy to eliminate countries with empty data
complete_final = complete_df.copy()

In [23]:
# Delete all rows with columns 'Migrant_Male_Population',"Migrant_Female_Population", "Total_Migrant_Population" that have value of 0
index_empty = complete_final.loc[(complete_final["Migrant_Male_Population"] == 0) | (complete_final["Migrant_Female_Population"] == 0) | (complete_final["Total_Migrant_Population"] == 0)].index
complete_final.drop(index_empty, inplace=True)
complete_final

Country_ID Country_ID_Alpha                   Country_Name  Year  \
0            108               BI                        Burundi  1990   
1            108               BI                        Burundi  1995   
2            108               BI                        Burundi  2000   
3            108               BI                        Burundi  2005   
4            108               BI                        Burundi  2010   
...          ...              ...                            ...   ...   
1633         876               WF     Wallis and Futuna Islands*  2000   
1634         876               WF     Wallis and Futuna Islands*  2005   
1635         876               WF     Wallis and Futuna Islands*  2010   
1636         876               WF     Wallis and Futuna Islands*  2015   
1637         876               WF     Wallis and Futuna Islands*  2020   

      Total_Migrant_Population  Migrant_Male_Population  \
0                       333110                   163267   
1                       254853                   124165   
2                       125628                    61094   
3                       192774                    94413   
4                       247321                   121762   
...                        ...                      ...   
1633                      2015                     1018   
1634                      2191                     1107   
1635                      2111                     1066   
1636                      2050                     1035   
1637                      2040                     1030   

      Migrant_Female_Population  
0                        169843  
1                        130688  
2                         64534  
3                         98361  
4                        125559  
...                         ...  
1633                        997  
1634                       1084  
1635                       1045  
1636                       1015  
1637                       1010  

[1595 rows x 7 columns]

In [22]:
complete_final.loc[(complete_final["Migrant_Male_Population"] == 0) | (complete_final["Migrant_Female_Population"] == 0) | (complete_final["Total_Migrant_Population"] == 0)].count()

Country_ID                   0
Country_ID_Alpha             0
Country_Name                 0
Year                         0
Total_Migrant_Population     0
Migrant_Male_Population      0
Migrant_Female_Population    0
dtype: int64

In [24]:
# Export the DataFrame
complete_final.to_csv("Migrant_Population_Final.csv", index=False)
files.download("Migrant_Population_Final.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

NOTE: Our dataset only includes data for 234 countries, territories or dependencies. The entities that are not on the dataset seem to be minor islands, including: Aland Islands, Antarctica, Bouvet Island, Christmas Island, Cocos (Keeling Islands), French Southern Territories, Guernsey, Heard Island and McDonald Islands, Jersey, Norfolk Island, Pitcairn, South Georgia and the South Sandwich Islands, Svalbard and Jan Mayen, United States Minor Outlying Islands.